src/renderers/shared/fiber/ReactFiberReconciler.js

```diff
@@ -170,7 +170,7 @@ module.exports = function<T, P, I, TI, PI, C, CX>(config : HostConfig<T, P, I, T
 
     deferredUpdates,
 
-    getPublicRootInstance(container : OpaqueNode) : (ReactComponent<any, any, any> | I | TI | null) {
+    getPublicRootInstance(container : OpaqueNode) : (ReactComponent<any, any, any> | PI | null) {
```

Comment:
how could I reproduce this scenario? None of the current tests seem to have an issue. I attempted to reproduce it with a root ref and the return value of `render` with a few component types.

```
describe('root level refs', () => {
  it('attaches and detaches root refs in stack', () => {
    assertForRenderer(require('ReactDOM'));
  });

  it('attaches and detaches root refs in fiber', () => {
    assertForRenderer(require('ReactDOMFiber'));
  });

  const assertForRenderer = (DOM) => {
    spyOn(console, 'error');
    var ref = jest.fn();
    var container = document.createElement('div');
    var result = DOM.render(<div ref={ref} />, container);
    expect(ref).toHaveBeenCalledTimes(1);
    expect(ref.mock.calls[0][0]).toBeInstanceOf(HTMLDivElement);
    expect(result).toBe(ref.mock.calls[0][0]);
    DOM.unmountComponentAtNode(container);
    expect(ref).toHaveBeenCalledTimes(2);
    expect(ref.mock.calls[1][0]).toBe(null);

    class Comp extends React.Component {
      render() {
        return <div>Comp</div>;
      }
    }

    ref = jest.fn();
    result = DOM.render(<Comp ref={ref}/>, container);

    expect(ref).toHaveBeenCalledTimes(1);
    expect(ref.mock.calls[0][0]).toBeInstanceOf(Comp);
    expect(result).toBe(ref.mock.calls[0][0]);
    DOM.unmountComponentAtNode(container);
    expect(ref).toHaveBeenCalledTimes(2);
    expect(ref.mock.calls[1][0]).toBe(null);
  };
});
```